### DISCUSSION TITLE
Use this extension along with "AutoHotKey2 Language Support" by thqby for syntax error highlighting

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [mark-wiemer/ahkpp#75] [Q&A] [answered]
So theres this other extension:
https://marketplace.visualstudio.com/items?itemName=thqby.vscode-autohotkey2-lsp

https://gitee.com/orz707/vscode-autohotkey2-lsp

which offers some really nice syntax highlighting, however its all in Chinese or something in that direction and I havent quite found a way of changing the language.

However when having both the vscode ahk language server extension from mark wiemer and this other extension it seems that when hovering some text its this other extension that displays. 

Does VS Code have some functionality that lets me choose which language server is responsible for what functionalities (or can I prioritize this extension somehow?).

Essentially what I wanna do is get the error syntax highlighting of this other extension, but use all the other features of Mark Wiemers ahk extension.

Thanks for reading :)

**Comment by fade2gray** [comment]
Don't know if this helps, but you can enable/disable similar extensions depending on which one you want to be in use. This can be done globally or per workspace.

![Capture](https://user-images.githubusercontent.com/1332009/111357941-b164ab80-8681-11eb-92dc-4b53b7c297d1.PNG)

**Reply by ItsCubeTime** [reply]
Thanks :) but I still want both extensions activated. I just want control over which extension does what since they provide the same functionality and kinda "step on eachothers toes".

**Reply by fade2gray** [reply]
Might not be the solution you want, but try...

1. Ensure both extensions are installed and enabled.
2. Open a new VSCode window,  and save the window as `workspace-1.code-workspace`.
3. Select `Disable (Workspace)` on one of the extensions using the `CogWheel/Manage menu` as shown above.
4. Open another new window and save it as  `workspace-2.code-workspace`.
5. Enable the disabled extension and disable the other extension using the `(Workspace)` option in all cases.

You may have to perform a `Reload` on one or both of the extensions but should find that both will work without interfering with each other in their separate Workspaces. 🙂

**Reply by FuPeiJiang** [reply]
add an option in `settings.json` to disable syntax highlighting ?

**Reply by ItsCubeTime** [reply]
> add an option in `settings.json` to disable syntax highlighting ?

Do you know what flags I should set?

**Reply by FuPeiJiang** [reply]
you can't right now, it has to be added in the code of this extension. should be easy to add this option, you just need to know wHeRe. and how

**Reply by ItsCubeTime** [reply]
Yeah okay I guess I will just live  with it then unless either Weimer or thqby adds support for it. Might try to reach out to the thqby about it

**Comment by mark-wiemer** [comment, marked as answer]
Hiya, sorry I never replied to this one. Thanks for calling out an extension with good syntax highlighting, I'll at it to my to-do list for May. As FuPeiJiang has mentioned, should be easy to add an option to disable AHK++ syntax highlighting so that it doesn't interfere with other syntax highlighting providers. No promises, but it's on the list :)

See [Add option to disable syntax highlighting · Issue #84 · mark-wiemer/vscode-autohotkey-plus-plus](https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/issues/84)

**Reply by ItsCubeTime** [reply]
> Hiya, sorry I never replied to this one. Thanks for calling out an extension with good syntax highlighting, I'll at it to my to-do list for May. As FuPeiJiang has mentioned, should be easy to add an option to disable AHK++ syntax highlighting so that it doesn't interfere with other syntax highlighting providers. No promises, but it's on the list :)
> 
> See [Add option to disable syntax highlighting · Issue #84 · mark-wiemer/vscode-autohotkey-plus-plus](https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/issues/84)

Hi there mark, I think you and FuPeiJiang misunderstood the situation exactly, its the other extension that takes precedence for the symbol hower information and such, hence Im reading this asian language rather than the output of your extension.

Im not familiar with the API but if its possible for you to give your extension some "higher priority" over other language server extensions that is more or less the fix that would be needed for this particular situation.

(Ability to disable features sounsds like a good idea still, it just wont be the answer to this issue - if the extension hosted on gitee had support for enabling/disabled each feature of the extension one by one this issue would never have been opened in the first place)

**Reply by mark-wiemer** [reply]
OK, I think I understand a bit better now. Can you refer me to what features of this other extension you want to keep? I see "Essentially what I wanna do is get the error syntax highlighting of this other extension", can you provide a screenshot with some AHK script and the behavior that you want?

I'm not sure about giving my extension higher priority, I'm hoping to be able to replicate the desired behavior in this extension so there's no need to use the other dependency, or disable the relevant feature from my extension so that there's no conflict.

**Reply by ItsCubeTime** [reply]
Hi, this is what I mean by syntax highlighting:

![image](https://user-images.githubusercontent.com/20190653/115855435-87926780-a42b-11eb-9f1c-33f764b69986.png)
(This is C++, but I think you get the point :P . The extension hosted on gitee seems to have some basic support for it)

**Reply by mark-wiemer** [reply]
Ref #469 :)

**Comment by mark-wiemer** [comment]
And yes, this other extension is in Chinese, according to a quick search

**Comment by mark-wiemer** [comment]
I just installed the extension directly from VS Code by searching `thqby.vscode-autohotkey2-lsp`, and its hovers are in English for me.

For example, the demo for AHK v2 code in the repo shows this:

```ahk
class Cls
{
    str := "string"
    num := 123
    static obj := { str: "string", num: 123 }
    propertyName[] { ; syntax error here
        set {
            this.propertyName_ := value
        }
        get {
            return this.propertyName_
        }
    }
    method() {
    }
    static staticMethod() {
    }
}
```

`propertyName[]` is underlined red with hover info "Empty [] not permitted". It looks like the Chinese is an issue with your version's localization, have you tried updating the version or installing directly from VS Code?

I know you still want my code's hover to take precedence, and I will look into that. I'm not sure how VS Code handles multiple hover providers, so it'll take some time.
